### ISSUE TITLE
Failure to load quarkus.properties

### ISSUE THREAD

**Issue by AhazIsraeli** [keycloak/keycloak#17398]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

dist/quarkus

### Describe the bug

Hello,
I've been trying to define keycloak database connection within the quarkus.properties file with the following variables:
quarkus.datasource.db-kind=mysql
quarkus.datasource.username=myuser
quarkus.datasource.password=mypassword
quarkus.datasource.jdbc.url=jdbc:mysql://myhost:3306/keycloak?characterEncoding=UTF-8&serverTimezone=UTC
The intention is eventually to use credentials provider for my instance.

I have placed this file under keycloak conf folder but it look like my keycloak instance ignores it altogether.
I am using keycloak 20.0.3.
Please advise

### Version

20.0.3

### Expected behavior

I am expecting keycloak to identify and relate to the definitions within my quarkus.properties file

### Actual behavior

keycloak ignores the definitions within my quarkus.properties file

### How to Reproduce?

Please see description

**Comment by iankko** [comment]
@AhazIsraeli Thanks for the report.

Can you clarify in which directory (relative to the main Keycloak on Quarkus dist config file `/opt/keycloak/conf/keycloak.conf`) is this `quarkus.properties` file located?

Yet, assuming under `ignores it altogether` you mean, the default H2 DB driver is loaded instead of the desired MySQL one. Is that correct?

**Comment by AhazIsraeli** [comment]
Hi,
I am working on Windows so in my case the folder path is C:\keycloak\20.0.3\conf
Your assumption is correct, the default H2 DB is being triggered.
Many thanks,
AHaz

**Comment by sschu** [comment]
You need to either start keycloak using `start-dev` so it takes build time options like `db-kind` into account or you have to create your own image using `kc.sh build' to pick up changed built time options. You can check https://www.keycloak.org/server/all-config to find out which options are build time and runtime options.

**Comment by AhazIsraeli** [comment]
Thank you Sebastian,
I tried all of your suggestions but without any success.
It looks like keycloak is ignoring any configuration appears on the quarkus.properties file no matter what I do.
I was wondering whether I could see something in the logs or increase the level of it so it may shed some light on this.
Best Regards,
Ahaz

**Comment by sschu** [comment]
Can you try if you get it to use your configured database by using environment variables for configuration?

**Comment by AhazIsraeli** [comment]
Will try happily.
What are the matching environment variables for the following?
quarkus.datasource.db-kind
quarkus.datasource.username
quarkus.datasource.password
quarkus.datasource.jdbc.url

**Comment by sschu** [comment]
Why are you using Quarkus properties for basic SQL connection properties? There are Keycloak-specific parameters for that?

**Comment by AhazIsraeli** [comment]
Hi,
This is because my aim eventually is to hide the password and for which case as far as I understood I'll need to use the quarkus.datasource.credentials-provider configuration
If you can advise me on another way to hide my password I'll be glad not using the quarkus.properties file
Best Regards,
Ahaz

**Comment by sschu** [comment]
Understood. You can find all the datasource config options including environment variable names here: https://quarkus.io/guides/datasource#jdbc-configuration First step is I would try to establish a basic connection without any additional stuff and then add the credential provider once this is working.

**Comment by AhazIsraeli** [comment]
Hi Sebastian,
This is exactly what I was trying to do.

I tried defining quarkus.properties with the following configurations:
quarkus.datasource.db-kind=mysql
quarkus.datasource.username=ahaz
quarkus.datasource.password=******
quarkus.datasource.jdbc.url=jdbc:mysql://<my-host>:3306/keycloak55?characterEncoding=UTF-8&serverTimezone=UTC

I tried defining environment variables as follows:
QUARKUS_DATASOURCE_DB_KIND=mysql
QUARKUS_DATASOURCE_JDBC_URL=jdbc:mysql://<my-host>:3306/keycloak?characterEncoding=UTF-8&serverTimezone=UTC
QUARKUS_DATASOURCE_PASSWORD=*****
QUARKUS_DATASOURCE_USERNAME=ahaz

But still after applying kc.bat build, kc.bat start (tried also kc.bat start-dev) these definitions were not accounted and keycloak created the default H2 database.

BTW: When I set the matching definitions within keycloak.conf the database connection is set as expected.

**Comment by sschu** [comment]
Looking at https://github.com/keycloak/keycloak/blob/f6a2b334d1e0652dcaba76cf8fccdc464a3171cd/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/DatabasePropertyMappers.java#L34, it might be that you cannot override properties that come from Keycloak. You might be able to set the ones that don't come from Keycloak. Could you try setting the basic DB settings using Keycloak parameters and only the additional ones you need via Quarkus? Just an idea, not sure it will work.

**Comment by AhazIsraeli** [comment]
Hi,
As explained above, I was able to connect to the database when the basic settings were set at keycloak.conf.
That is, the following attributes were set: db, db-username, db-password, db-url.
What do you mean by saying: 'only additional ones'? is this the quarkus.datasource.credentials-provider? If this is the case I never tried it as I want the basic database connection to work successfully.
BTW: when trying to use the settings from quarkus.properties file the ones from keycloak.comf were deleted.
Best Regards,
Ahaz

**Comment by sschu** [comment]
Exactly, I meant only specifying `quarkus.datasource.credentials-provider`as Quarkus properties but the rest from Keycloak.

**Comment by AhazIsraeli** [comment]
Hi again,
Defining only quarkus.datasource.credentials-provider within the quarkus.properties file seems to did the trick as I think keycloak is referring to the value I specify there.

Now, with that being said, when I set the following definitions.
quarkus.datasource.credentials-provider = custom
quarkus.datasource.credentials-provider-name = my-credentials-provider

I always get the following error:
java.lang.RuntimeException: unable to find credentials provider of type my-credentials-provider

I have created a JAR file containing an implementation of io.quarkus.credentials.CredentialsProvider interface and I have placed it under '<keycloak-home>\lib\lib'

What am I missing?

**Comment by sschu** [comment]
I have no experience with custom credentials-providers. Have you checked https://quarkus.io/guides/credentials-provider?

**Comment by AhazIsraeli** [comment]
Yes I did. Opening this item is my final solution after essentially giving up.
Believe me, I have tried everything I could think of and read through several documents and ideas.

**Comment by AhazIsraeli** [comment]
Hello,
I was wondering whether anyone has a solution for me?
How can I implement a quarkus custom credentials provider and deploy it successfully within my keycloak installation.
Many thanks in advance,
Ahaz

**Comment by vmuzikar** [comment]
@AhazIsraeli Thanks for the report.

While we do not support integration Quarkus Credentials provider, we recently [added support for SmallRye Keystore](https://github.com/keycloak/keycloak/pull/20375). It will be included in the next major release: Keycloak 22. But you can already give it a try in our nightly releases!

Please note that we do not support configuring raw Quarkus options for which we have a Keycloak wrapper (like `quarkus.datasource.username`) either.

I'm going to close this as it is not a bug. Feel free to open an RFE if the SmallRye Keystore doesn't cover your requirements. Thank you.
